pkg/network/blockqueue.go

```diff
@@ -39,67 +42,93 @@ func newBlockQueue(capacity int, bc blockchainer.Blockqueuer, log *zap.Logger, r
 }
 
 func (bq *blockQueue) run() {
+	var lastHeight = bq.chain.BlockHeight()
 	for {
 		_, ok := <-bq.checkBlocks
 		if !ok {
 			break
 		}
 		for {
-			item := bq.queue.Peek()
-			if item == nil {
+			h := bq.chain.BlockHeight()
+			pos := int(h+1) % blockCacheSize
+			bq.queueLock.Lock()
+			b := bq.queue[pos]
+			// The chain moved forward using blocks from other sources (consensus).
+			for i := lastHeight; i < h; i++ {
+				old := int(i+1) % blockCacheSize
+				if bq.queue[old] != nil && bq.queue[old].Index == i {
+					bq.len--
+					bq.queue[old] = nil
+				}
+			}
+			bq.queueLock.Unlock()
+			lastHeight = h
+			if b == nil {
 				break
 			}
-			minblock := item.(*block.Block)
-			if minblock.Index <= bq.chain.BlockHeight()+1 {
-				_, _ = bq.queue.Get(1)
-				updateBlockQueueLenMetric(bq.length())
-				if minblock.Index == bq.chain.BlockHeight()+1 {
-					err := bq.chain.AddBlock(minblock)
-					if err != nil {
-						// The block might already be added by consensus.
-						if bq.chain.BlockHeight() < minblock.Index {
-							bq.log.Warn("blockQueue: failed adding block into the blockchain",
-								zap.String("error", err.Error()),
-								zap.Uint32("blockHeight", bq.chain.BlockHeight()),
-								zap.Uint32("nextIndex", minblock.Index))
-						}
-					} else if bq.relayF != nil {
-						bq.relayF(minblock)
-					}
+
+			err := bq.chain.AddBlock(b)
+			if err != nil {
+				// The block might already be added by consensus.
+				if bq.chain.BlockHeight() < b.Index {
+					bq.log.Warn("blockQueue: failed adding block into the blockchain",
+						zap.String("error", err.Error()),
+						zap.Uint32("blockHeight", bq.chain.BlockHeight()),
+						zap.Uint32("nextIndex", b.Index))
 				}
-			} else {
-				break
+			} else if bq.relayF != nil {
+				bq.relayF(b)
 			}
+			bq.queueLock.Lock()
+			bq.len--
+			l := bq.len
+			if bq.queue[pos] == b {
+				bq.queue[pos] = nil
+			}
+			bq.queueLock.Unlock()
+			updateBlockQueueLenMetric(l)
 		}
 	}
 }
 
 func (bq *blockQueue) putBlock(block *block.Block) error {
 	h := bq.chain.BlockHeight()
+	bq.queueLock.Lock()
 	if block.Index <= h || h+blockCacheSize < block.Index {
 		// can easily happen when fetching the same blocks from
 		// different peers, thus not considered as error
+		bq.queueLock.Unlock()
 		return nil
 	}
-	err := bq.queue.Put(block)
+	pos := block.Index % blockCacheSize
+	// If we already have it, keep the old block, throw away new one.
+	if bq.queue[pos] == nil || bq.queue[pos].Index < block.Index {
```

Comment:
So if `bq.queue[pos].Index < block.Index` then you replace the old block with the new one. But what if the old block still wasn't added to the chain and we still need it?

Reply:
I don't think it's possible with `if block.Index <= h || h+blockCacheSize < block.Index {` condition above. We know `block.Index` is within `blockCacheSize` from current height, so if there is anything in the slot, it's outdated.

Reply:
Missed that, so it's ok.